src/mlpack/methods/ann/rbm/rbm_impl.hpp

```diff
@@ -0,0 +1,351 @@
+>
+template<typename OptimizerType>
+void RBM<InitializationRuleType, DataType, PolicyType>::Train(
+    OptimizerType& optimizer)
+{
+  if (!reset)
+  {
+    Reset();
+  }
+
+  optimizer.Optimize(*this, parameter);
+}
+
+template<
+  typename InitializationRuleType,
+  typename DataType,
+  typename PolicyType
+>
+template<typename Policy>
+typename std::enable_if<std::is_same<Policy, BinaryRBM>::value, double>::type
+RBM<InitializationRuleType, DataType, PolicyType>::FreeEnergy(
+    arma::Mat<ElemType>&& input)
+{
+  preActivation = (weight * input);
+  preActivation.each_col() += hiddenBias;
+  preActivation = arma::log(1 + arma::trunc_exp(preActivation));
+  return -(arma::accu(preActivation) + arma::dot(input, arma::repmat(
+      visibleBias, 1, input.n_cols)));
+}
+
+template<
+  typename InitializationRuleType,
+  typename DataType,
+  typename PolicyType
+>
+template<typename Policy>
+typename std::enable_if<std::is_same<Policy, BinaryRBM>::value, void>::type
+RBM<InitializationRuleType, DataType, PolicyType>::Phase(
+    DataType&& input,
+    DataType&& gradient)
+{
+  DataType weightGrad = DataType(gradient.memptr(),
+      hiddenSize, visibleSize, false, false);
+
+  DataType hiddenBiasGrad = DataType(gradient.memptr() + weightGrad.n_elem,
+      hiddenSize, 1, false, false);
+
+  DataType visibleBiasGrad = DataType(gradient.memptr() + weightGrad.n_elem +
+      hiddenBiasGrad.n_elem, visibleSize, 1, false, false);
+
+  HiddenMean(std::move(input), std::move(hiddenBiasGrad));
+  weightGrad = hiddenBiasGrad * input.t();
+  visibleBiasGrad = input;
+}
+
+template<
+  typename InitializationRuleType,
+  typename DataType,
+  typename PolicyType
+>
+double RBM<InitializationRuleType, DataType, PolicyType>::Evaluate(
+    const arma::Mat<ElemType>& /* parameters*/,
+    const size_t i,
+    const size_t batchSize)
+{
+  Gibbs(std::move(predictors.cols(i, i + batchSize - 1)),
+      std::move(negativeSamples));
+  return std::fabs(FreeEnergy(std::move(predictors.cols(i,
+      i + batchSize - 1))) - FreeEnergy(std::move(negativeSamples)));
+}
+
+template<
+  typename InitializationRuleType,
+  typename DataType,
+  typename PolicyType
+>
+template<typename Policy>
+typename std::enable_if<std::is_same<Policy, BinaryRBM>::value, void>::type
+RBM<InitializationRuleType, DataType, PolicyType>::SampleHidden(
+    arma::Mat<ElemType>&& input,
+    arma::Mat<ElemType>&& output)
+{
+  HiddenMean(std::move(input), std::move(output));
+
+  for (size_t i = 0; i < output.n_elem; i++)
+  {
+    output(i) = math::RandBernoulli(output(i));
+  }
+}
+
+template<
+  typename InitializationRuleType,
+  typename DataType,
+  typename PolicyType
+>
+template<typename Policy>
+typename std::enable_if<std::is_same<Policy, BinaryRBM>::value, void>::type
+RBM<InitializationRuleType, DataType, PolicyType>::SampleVisible(
+    arma::Mat<ElemType>&& input,
+    arma::Mat<ElemType>&& output)
+{
+  VisibleMean(std::move(input), std::move(output));
```

Comment:
Do you think, we could just use a reference for the ``input`` and ``output`` term?

Reply:
Hmm, to do that, we'll have to change the signatures of all the intermediate functions to use references. How should we proceed here?